library/footer.cpp

```diff
@@ -0,0 +1,289 @@
+		return err;
+	}
+
+	memset(csum, 0, sizeof(csum));
+
+	off = wc->ctl_data_offset + sizeof(struct eblob_disk_control);
+	err = sha512_file(wc->data_fd, off, wc->total_data_size, csum);
+	if (err) {
+		eblob_log(b->cfg.log, EBLOB_LOG_ERROR, "blob i%d: %s: %s: sha512_file failed: err: %d\n",
+		          wc->bctl->index, eblob_dump_id(key->id), __func__, err);
+		return err;
+	}
+
+	if (memcmp(csum, f.csum, sizeof(csum))) {
+		eblob_log(b->cfg.log, EBLOB_LOG_ERROR, "blob i%d: %s: %s: checksum mismatch: err: %d\n",
+		          wc->bctl->index, eblob_dump_id(key->id), __func__, err);
+		return -EILSEQ;
+	}
+
+	eblob_log(b->cfg.log, EBLOB_LOG_NOTICE, "blob: i%d: %s: %s: checksum verified\n",
+	          wc->index, eblob_dump_id(key->id), __func__);
+
+	return 0;
+}
+
+
+/*
+ * eblob_verify_crc32() - verifies checksum of entry pointed by @wc by comparing crc32 of record's data chunks with footer.
+ * It will checks only chunks that intersect @wc->offset and @wc->size.
+ *
+ * Returns negative error value or zero on success.
+ */
+static int eblob_verify_crc32(struct eblob_backend *b, struct eblob_key *key, struct eblob_write_control *wc) {
+	int err = 0;
+	uint64_t footers_offset = 0,
+	         footers_size = 0;
+
+	std::vector<uint32_t> calc_footers, check_footers;
+
+	err = eblob_chunked_crc32(b, key, wc, wc->offset, wc->size, calc_footers, footers_offset);
+	if (err) {
+		eblob_log(b->cfg.log, EBLOB_LOG_ERROR, "blob i%d: %s: %s: eblob_chunked_crc32: failed: fd: %d, size: %"PRIu64
+		          ", offset: %" PRIu64 "\n",
+		          wc->bctl->index, eblob_dump_id(key->id), __func__, wc->data_fd, footers_size, footers_offset);
+		return err;
+	}
+
+	footers_size = calc_footers.size() * sizeof(calc_footers.front());
+
+	check_footers.resize(calc_footers.size(), 0);
+	err = __eblob_read_ll(wc->data_fd, check_footers.data(), footers_size, footers_offset);
+	if (err) {
+		eblob_log(b->cfg.log, EBLOB_LOG_ERROR, "blob i%d: %s: %s: failed to read footer: fd: %d, size: %"PRIu64
+		          ", offset: %" PRIu64 "\n",
+		          wc->index, eblob_dump_id(key->id), __func__, wc->data_fd, footers_size, footers_offset);
+		return err;
+	}
+
+	if (memcmp(calc_footers.data(), check_footers.data(), footers_size)) {
+		eblob_log(b->cfg.log, EBLOB_LOG_DEBUG, "blob i%d: %s: %s: checksum mismatch: footers_size: %" PRIu64
+		          ", footers_count: %" PRIu64"\n",
+		          wc->index, eblob_dump_id(key->id), __func__, footers_size, calc_footers.size());
+		return -EILSEQ;
+	}
+
+	eblob_log(b->cfg.log, EBLOB_LOG_NOTICE, "blob: i%d: %s: %s: checksum verified\n",
+	          wc->index, eblob_dump_id(key->id), __func__);
+
+	return 0;
+}
+
+int eblob_verify_checksum(struct eblob_backend *b, struct eblob_key *key, struct eblob_write_control *wc) {
+	if (b->cfg.blob_flags & EBLOB_NO_FOOTER ||
+	    wc->flags & BLOB_DISK_CTL_NOCSUM)
+		return 0;
+
+	if (wc->total_size < wc->total_data_size)
```

Comment:
`eblob_verify_checksum()` is API function and can be called by client. It should include sizeof(eblob_disk_control). I will add it and another sanity checks to verification functions.